finetune/utils.py

```diff
@@ -376,23 +395,58 @@ def indico_to_finetune_sequence(texts, labels=None, none_value=config.PAD_TOKEN)
         labels = [[]] * len(texts)
 
     for text, label_seq in zip(texts, labels):
+        tokens = NLP(text)
+        token_starts = [token.idx for token in tokens]
+        token_ends = [token.idx + len(token.text) for token in tokens]
+        n_tokens = len(tokens)
+
+        label_seq = sorted(label_seq, key=lambda x: x["start"])
         last_loc = 0
         doc_subseqs = []
         doc_labels = []
-        for annotation in label_seq:
+        for i, annotation in enumerate(label_seq):
             start = annotation["start"]
             end = annotation["end"]
             label = annotation["label"]
-            if start != last_loc:
+            if not subtoken_labels:
+                if label != none_value:
+                    # round to nearest token
+                    while start > 0 and start not in token_starts:
+                        start -= 1
+                    while end < len(text) and end not in token_ends:
+                        end += 1
+
+            if start > last_loc:
                 doc_subseqs.append(text[last_loc:start])
-                doc_labels.append(none_value)
+                if multi_label:
+                    doc_labels.append([none_value])
+                else:
+                    doc_labels.append(none_value)
+
+            if start < last_loc:
+                if not multi_label:
+                    raise ValueError("Overlapping annotations requires the multi-label model")
+                else:
+                    prev_end = label_seq[i - 1]["end"]
```

Comment:
why are these set to the same thing?

Reply:
Tests to write:
Overlapping
Label contained within another
More than 2 labels